src/Schema/Directives/Fields/MiddlewareDirective.php

```diff
@@ -27,52 +51,49 @@ public function name()
      *
      * @return FieldValue
      */
-    public function handleField(FieldValue $value, \Closure $next)
+    public function handleField(FieldValue $value, \Closure $next): FieldValue
     {
-        $checks = $this->getChecks($value);
-
-        if ($checks) {
-            $middlewareRegistry = resolve(MiddlewareRegistry::class);
-
-            if ('Query' === $value->getNodeName()) {
-                $middlewareRegistry->registerQuery(
-                    $value->getFieldName(),
-                    $checks
-                );
-            } elseif ('Mutation' === $value->getNodeName()) {
-                $middlewareRegistry->registerMutation(
-                    $value->getFieldName(),
-                    $checks
-                );
-            }
-        }
-
-        return $next($value);
-    }
-
-    /**
-     * Get middleware checks.
-     *
-     * @param FieldValue $value
-     *
-     * @return array|null
-     */
-    protected function getChecks(FieldValue $value)
-    {
-        if (! in_array($value->getNodeName(), ['Mutation', 'Query'])) {
-            return null;
-        }
-
-        $checks = $this->directiveArgValue('checks');
+        /** @var Router $router */
+        $router = resolve('router');
+        $middleware = $router->getMiddleware();
+        $middlewareGroups = $router->getMiddlewareGroups();
 
-        if (! $checks) {
-            return null;
-        }
+        $middleware = collect($this->directiveArgValue('checks'))
+            ->map(function ($name) use ($middleware, $middlewareGroups){
+                return (array) MiddlewareNameResolver::resolve($name, $middleware, $middlewareGroups);
+            })
+            ->flatten();
 
-        if (is_string($checks)) {
-            $checks = [$checks];
-        }
+        $resolver = $value->getResolver();
 
-        return $checks;
+        return $next(
+            $value->setResolver(
+                function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo) use ($resolver, $middleware) {
+                    dump($resolveInfo->fieldName, $middleware);
```

Comment:
Note: These dumps break json output (workaround: use dump-server)

Reply:
Thanks for the review, this is still a WIP, i will remove the dumps once it is finished. Any opinions on the rest?

Reply:
I checked this out to benchmark and test it, unfortunately my relatively unperformant schema does not work with latest master or this branch, so I'm still figuring out whats causing the issues.
I will give some more feedback later